finetuner/tuner/evaluation.py

```diff
@@ -0,0 +1,161 @@
+from typing import TYPE_CHECKING, Optional, List, Dict, Tuple
+from jina import Document, DocumentArray
+from jina.math.evaluation import (
+    r_precision,
+    precision_at_k,
+    recall_at_k,
+    f1_score_at_k,
+    hit_at_k,
+    average_precision,
+    reciprocal_rank,
+    dcg_at_k,
+    ndcg_at_k,
+)
+
+from .. import __default_tag_key__
+from ..embedding import embed
+
+if TYPE_CHECKING:
+    from ..helper import AnyDNN, DocumentSequence
+
+
+METRICS = {
+    "r_precision": r_precision,
+    "precision": precision_at_k,
+    "recall": recall_at_k,
+    "f1score": f1_score_at_k,
+    "average_precision": average_precision,
+    "hit": hit_at_k,
+    "reciprocal_rank": reciprocal_rank,
+    "dcg": dcg_at_k,
+    "ndcg": ndcg_at_k,
+}
+
+
+class Evaluator:
+    def __init__(
+        self,
+        eval_data: 'DocumentSequence',
+        catalog: 'DocumentSequence',
+        embed_model: Optional['AnyDNN'] = None,
+        distance: str = 'cosine',
+        limit: int = 20,
+    ):
+        """
+        Build an Evaluator object that can be used to evaluate an embedding model on a retrieval task
+
+        :param eval_data: A sequence of documents. Each document should contain ground truth matches under
+            ``doc.matches`` and relevance scores under ``doc.tags[__default_tag_key__]['label']``
+        :param catalog: A sequence of documents, against which the eval docs will be matched.
+        :param embed_model: The embedding model to use, in order to extract document representations.
+        :param distance: The type of distance to use when matching docs, avalilable options are
+            ``"cosine"``, ``"euclidean"`` and ``"sqeuclidean"``
+        :param limit: Limit the number of results during matching.
+        :return: None.
+        """
+        self._embed_model = embed_model
+        self._eval_data = eval_data
+        self._catalog = catalog
+        self._summary_docs = self._parse_eval_docs()
+        self._distance = distance
+        self._limit = limit
+
+    @staticmethod
+    def _doc_to_relevance(doc: Document) -> Tuple[List[int], int]:
+        """
+        Convert a Jina document to a relevance representation
+        :param doc: A Jina document. The matched document identifiers, predicted by the search system, are expected to
+            be under ``doc.matches``. The expected documents identifiers along with the relevance scores, given by the
+            user as matching ground truth, should be under ``doc.tags[finetuner.__default_tag_key__]['targets']``.
```

Comment:
if now label is `__finetuner_label__`, how does it affects this?

Reply:
in the internal docs that the evaluator holds (`self._summary_docs`) ground truth matches will be stored in `doc.tags["finetuner_label"]["targets"]`